src/MassTransit/Context/InMemoryOutboxMessageSchedulerContext.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright 2007-2016 Chris Patterson, Dru Sellers, Travis Smith, et. al.
+//  
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the 
+// License at 
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0 
+// 
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR 
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the 
+// specific language governing permissions and limitations under the License.
+namespace MassTransit.Context
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using GreenPipes;
+    using Scheduling;
+
+
+    public class InMemoryOutboxMessageSchedulerContext :
+        MessageSchedulerContext
+    {
+        readonly MessageSchedulerContext _context;
+        readonly object _listLock = new object();
+        readonly List<ScheduledMessage> _scheduledMessages;
+
+        public InMemoryOutboxMessageSchedulerContext(MessageSchedulerContext context)
+        {
+            _context = context ?? throw new ArgumentNullException(nameof(context));
+
+            _scheduledMessages = new List<ScheduledMessage>();
+        }
+
+        void AddScheduledMessage(ScheduledMessage scheduledMessage)
+        {
+            lock (_listLock)
+                _scheduledMessages.Add(scheduledMessage);
+        }
+
+        public Task CancelAllScheduledMessages()
+        {
+            ScheduledMessage[] scheduledMessages;
+
+            lock (_listLock)
```

Comment:
`lock` is not working with `Task`s, you have to think about `Semaphore` or `SemaphoreSlim`

Reply:
Really? I'm using the same technique as [here](https://github.com/MassTransit/MassTransit/blob/6ab652e21b8d3276f59b52ae1f6072922d8f6144/src/MassTransit/Context/InMemoryOutboxConsumeContext.cs#L34).

Reply:
just check this one :)
```csharp
private static readonly object Lock = new object();
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            for (int i = 0; i < 3; i++)
            {
                var i1 = i;
                Task.Run(() =>
                {
                    lock (Lock)
                    {
                        Console.WriteLine(i1);
                        return Task.Delay(TimeSpan.FromSeconds(30));
                    }
                });
            }

            Console.ReadLine();
        }
    }
```

Reply:
actually, your code will be ran sync. it is safe to use it here, sorry

Reply:
thanks for being a second set of eyes on the PR. Appreciate it!